README.md
# Web Developer test exercise

## Exercise

> _In a nutshell a simple many-to-many relationship modeling with some primitive CRUD and UI_

### Context

We have a web application where our admins can see/review/manage users.

### Challenge

Our customer-success department is missing functionality to group users based on their skills.

### User story

- As CS-manager, I want to **attach skills to users**, so I can easily **filter the list of users by skill**.

### Acceptance criteria (happy flow)

1. I choose a user
2. I attach skill "javascript" to the user
3. I filter the list of users by skill "javascript"
4. I see result 1 row with the user from 2nd step

## Setup & run project

### Backend

> _Assumption is that you working under ubuntu and have set up python3.8_
> 
> If you have difficulties to write backend in python, no problem, you can as well use any nodejs backend framework for that. Important part here is to complete the challenge ;)

```shell
cd backend

# Install virtual environment
python3.8 -m venv .venv
source .venv/bin/activate

# Install dependencies
pip install -r requirements.txt

# Run
flask run --reload
```

### Frontend

> _Assumption is that you already installed Node 12+_

```shell
cd frontend

# Install dependencies
npm install

# Run
npm start
```



backend/CHANGELOG.md
### Added
- Skills model
- User routes
- Skill routes
- many-to-many association

### Changed
- Updated Flask to the latest version

### Test
- Written tests using unittest


backend/tests/test.py
import unittest
import json
from app import app

# Test cases


class UserAPITestCase(unittest.TestCase):
    def setUp(self):
        app.config.from_object(self)
        self.client = app.test_client()

    def test_post_users(self):
        response = self.client.post('/users', content_type='application/json')
        self.assertEqual(response.status_code, 201)

        response = self.client.get('/users')
        self.assertEqual(response.status_code, 200)

        data = json.loads(response.data)
        self.assertGreater(len(data['items']), 0)

    def test_post_skill(self):
        data = {'name': 'Python'}
        response = self.client.post('/skill', data=json.dumps(data), content_type='application/json')
        self.assertEqual(response.status_code, 201)
        data = json.loads(response.data)
        self.assertEqual(data['skill']['name'], 'Python')

        response = self.client.get('/skill')
        self.assertEqual(response.status_code, 200)
        data = json.loads(response.data)
        print(data)
        self.assertGreater(len(data['items']), 0)

    def test_add_skill_to_user(self):
        skill = {'name': 'Java'}
        user = {'name': 'X-T'}
        # create user
        response = self.client.post('/user', data=json.dumps(user), content_type='application/json')
        self.assertEqual(response.status_code, 201)

        # add skill
        response = self.client.post('/user/1/skill', data=json.dumps(skill), content_type='application/json')
        self.assertEqual(response.status_code, 201)
        data = json.loads(response.data)
        self.assertGreater(len(data['user']['skills']), 0)


if __name__ == '__main__':
    unittest.main()


backend/models.py
from typing import List, ForwardRef, Dict

from flask_sqlalchemy import SQLAlchemy
from pydantic import BaseModel

db = SQLAlchemy()


"""
FOR SKILLS
"""


class Skill(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(100), nullable=False)

    def __repr__(self):
        return "<Sill %r>" % self.name


class SkillSchema(BaseModel):
    id: int
    name: str

    class Config:
        orm_mode = True


class SkillCollectionResponse(BaseModel):
    items: List[SkillSchema]


class SkillResponse(BaseModel):
    skill: SkillSchema


'''
FOR USERS
'''


class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(80), unique=True, nullable=False)
    skills = db.relationship('Skill', secondary='user_skills', lazy="subquery",  backref=db.backref("users")
                             , cascade="all,delete")

    def __repr__(self):
        return "<User %r>" % self.name


class UserSchema(BaseModel):
    id: int
    name: str
    skills: List[SkillSchema]

    class Config:
        orm_mode = True


class UserCollectionResponse(BaseModel):
    items: List[UserSchema]


class UserResponse(BaseModel):
    user: UserSchema


class UserSkills(db.Model):
    """
    for many-to-many association,
    onDelete CASCADE
    """
    user_id = db.Column(db.Integer, db.ForeignKey('user.id', ondelete='CASCADE'), primary_key=True)
    skill_id = db.Column(db.Integer, db.ForeignKey('skill.id',  ondelete='CASCADE'), primary_key=True)


backend/app.py
from faker import Faker
from flask import Flask, request, jsonify
from flask_cors import CORS

from models import db, User, UserCollectionResponse, UserResponse, Skill, SkillCollectionResponse, SkillResponse

fake = Faker()


def create_app():
    _app = Flask(__name__)
    _app.config["SQLALCHEMY_DATABASE_URI"] = "sqlite:///:memory:"
    _app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
    db.init_app(_app)
    with _app.app_context():
        db.drop_all()
        db.create_all()
    return _app


app = create_app()

CORS(app)


'''
ROUTES FOR BULK ACTIONS ON USER
'''


@app.route("/users", methods=["GET"])
def users():
    with app.app_context():
        skill_to_filter = request.args.get("skill")
        if skill_to_filter:
            filtered_users = User.query.filter(User.skills.any(Skill.id == skill_to_filter)).all()
            return UserCollectionResponse(items=filtered_users).json()
        results = User.query.all()
    return UserCollectionResponse(items=results).json(), 200


@app.route("/users", methods=["POST"])
def create_users_batch():
    with app.app_context():
        '''
        To Enforce Foreign Keys
        '''
        db.session.execute('pragma foreign_keys=on')
        for x in range(10):
            db.session.add(User(name=fake.name()))
        db.session.commit()
    return "Users created", 201


@app.route("/users", methods=["DELETE"])
def delete_all_users():
    with app.app_context():
        db.session.query(User).delete()
        db.session.commit()
    return "Users deleted", 200


'''
ROUTES FOR USER
'''


@app.route('/user/<int:user_id>', methods=['GET'])
def user(user_id):
    """
    GET a user with user_id
    """
    existing_user = User.query.get(user_id)
    if not existing_user:
        return "User not found", 404
    return UserResponse(user=existing_user).json()


@app.route('/user/<int:user_id>', methods=['DELETE'])
def delete_user(user_id):
    """
    DELETE a single user with user_id
    """
    user_to_delete = User.query.get(user_id)
    if user_to_delete:
        db.session.delete(user_to_delete)
    db.session.commit()
    return "User deleted", 201


@app.route('/user', methods=['POST'])
def create_user():
    """
    CREATE a single user with a specific name in request
    """
    data = request.get_json()
    db.session.add(User(name=data['name']))
    db.session.commit()
    return "Users created", 201


@app.route("/user/<int:user_id>/skill", methods=["POST"])
def add_skill_to_user(user_id):
    """
    This function adds skill to user if skill exists in DB
    If not, add skill to db and then add to user
    """
    m_user = User.query.get(user_id)
    if not m_user:
        return "User not found", 404
    data = request.get_json()
    skill_name = data.get("name")
    skill = Skill.query.filter_by(name=skill_name).first()
    if skill is None:
        skill = Skill(name=skill_name)
        db.session.add(skill)
        db.session.commit()
    elif skill in m_user.skills:
        return "Skill already added to user", 400
    m_user.skills.append(skill)
    db.session.commit()
    return UserResponse(user=m_user).json(), 201


@app.route("/user/<int:user_id>/skill/<int:skill_id>", methods=["DELETE"])
def delete_skill_from_user(user_id, skill_id):
    """
    DELETE a skill with skill_id from a user with user_id
    """
    existing_user = User.query.get(user_id)
    if existing_user is None:
        return 'User not found', 404
    existing_skill = Skill.query.get(skill_id)
    if existing_skill is None:
        return 'Skill not found', 404
    existing_user.skills.remove(existing_skill)
    db.session.commit()
    return UserResponse(user=existing_user).json(), 200


'''
ROUTES FOR SKILL
'''


@app.route("/skill", methods=["GET"])
def all_skills():
    """GET all saved skills from DB"""
    with app.app_context():
        results = Skill.query.all()
    return SkillCollectionResponse(items=results).json(), 200


@app.route("/skill/<int:skill_id>", methods=["GET"])
def skill(skill_id):
    existing_skill = Skill.query.get(skill_id)
    if not existing_skill:
        return "Skill not found", 404
    db.session.commit()
    return SkillResponse(skill=existing_skill).json(), 200


@app.route("/skill", methods=["POST"])
def create_skill():
    data = request.get_json()
    skill_name = data.get("name")

    new_skill = Skill(name=skill_name)
    db.session.add(new_skill)
    db.session.commit()
    return SkillResponse(skill=new_skill).json(), 201


@app.route("/skill/<int:skill_id>", methods=["PUT"])
def update_skill(skill_id):
    """UPDATE a skill name with a new name from request"""
    skill_to_update = Skill.query.get(skill_id)
    if not skill_to_update:
        return "Skill not found", 404
    data = request.get_json()
    skill_to_update.name = data['name']
    db.session.commit()
    return SkillResponse(skill=skill_to_update).json(), 200


@app.route("/skill/<int:skill_id>", methods=["DELETE"])
def delete_skill(skill_id):
    skill_to_delete = Skill.query.get(skill_id)
    if not skill_to_delete:
        return "Skill not found", 404

    db.session.delete(skill_to_delete)
    db.session.commit()

    return "Skill deleted successfully", 200


if __name__ == "__main__":
    app.run()
